### PR TITLE
adding charting function for Sankey diagram

### PR DESCRIPTION
**Pull request by agaperste**
Starting to add functions for more charting capabilities, starting with making a Sankey diagram.

### CODE REVIEW

File: tests/unit/test_viz_sankey.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+from unittest.mock import patch
+import pandas as pd
+from dune_client.viz.graphs import create_sankey
+
+
+class TestCreateSankey(unittest.TestCase):
+
+    # Setting up a dataframe for testing
+    def setUp(self):
+        self.df = pd.DataFrame(
```
Comment (bh2smith) [inline tests/unit/test_viz_sankey.py:11]: Could this test be made much smaller and still demonstrate the point? Like with two tokens?

File: dune_client/viz/graphs.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+"""
+Functions you can call to make different graphs
+"""
+
+import plotly.graph_objects as go
+import colorlover as cl
+import pandas as pd
```
Comment (bh2smith) [inline dune_client/viz/graphs.py:7]: You're going to need to add all of these packages to the requirements/prod.txt file and also in setup.cfg:

https://github.com/duneanalytics/dune-client/blob/b510c974f2e826a0f6e72392b4988515550772a2/setup.cfg#L22

File: dune_client/viz/graphs.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+# function to create Sankey diagram
+def create_sankey(
+    query_result: pd.DataFrame,
+    predefined_colors: dict,
+    columns: dict,
+    viz_config: dict,
```
Comment (bh2smith) [inline dune_client/viz/graphs.py:14]: the test suggests that these could be 
```suggestion
    predefined_colors: dict[str, str],
    columns: dict[str, str],
    viz_config: dict[str, int | float],
```

Although it also kinda looks like columns could be `list[str]`. 

Note that earlier python versions might require `Union[int, float]` (if you decide to go with this suggestion).

File: tests/unit/test_viz_sankey.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            "USDT": "rgb(0, 143, 142)",
+            "WETH": "rgb(144, 144, 144)",
+            "WBTC": "rgb(247, 150, 38)",
+            "COMP": "rgb(32, 217, 152)",
+            "DAI": "rgb(254, 175, 48)",
+            "MKR": "rgb(38, 173, 158)",
+            "UNI": "rgb(255, 21, 126)",
+        }
+
+        self.columns = {"source": "source", "target": "target", "value": "value"}
```
Comment (bh2smith) [inline tests/unit/test_viz_sankey.py:96]: Why are the columns mappings of themselves to themselves? Is there an example where they aren't just the same values mapping to themselves or does this even need to be a dict?

**Review by bh2smith — COMMENTED** [review]
Looks interesting! Left a few suggestions and notes that might help pass CI. It might also be reasonable to simply add these new dependencies to the dev-dependencies so that users can add them if they decide to use this feature (currently pandas is setup like that).

File: tests/unit/test_viz_sankey.py (hunk 1 above)
Comment (agaperste) [reply tests/unit/test_viz_sankey.py:11]: good point shortened!

File: dune_client/viz/graphs.py (hunk 2 above)
Comment (agaperste) [reply dune_client/viz/graphs.py:8]: ah thank you! I was so confused why the build kept failing here! decided to go with the dev route!

File: tests/unit/test_viz_sankey.py (hunk 4 above)
Comment (agaperste) [reply tests/unit/test_viz_sankey.py:24]: I was thinking of giving folks the option to name their columns differently, which ultimately has to be mapped to be source, target, value.

**Comment by agaperste** [comment]
@bh2smith Thanks for giving me all the pointers! When you get a chance, can you help me with this error from black? I ran stuff in local and the formatting is okay, but on GitHub it's still failing 
Screen Shot 2023-09-29 at 7 24 14 PM

File: tests/unit/test_viz_sankey.py (hunk 4 above)
Comment (bh2smith) [reply tests/unit/test_viz_sankey.py:24]: Perhaps this fact can be included somewhere in the doc strings. And also here in the test you could use targets like "renamed_column"

File: tests/unit/test_viz_sankey.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        df_without_target = self.df.drop(columns=["target"])
+        with self.assertRaises(ValueError):
+            create_sankey(
+                df_without_target, self.predefined_colors, self.columns, self.viz_config
+            )
+
+    def test_value_column_not_numeric(self):
+        # Change the 'value' column to a non-numeric type
+        df_with_str_values = self.df.copy()
+        df_with_str_values["value"] = ["10"] * len(df_with_str_values)
```
Comment (bh2smith) [inline tests/unit/test_viz_sankey.py:44]: Might be easier here to just ... is there some way to ensure types on the Dataframe by definition? 
```suggestion
        df_with_str_values["value"] = ["10", "11"]
```

File: dune_client/viz/graphs.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+def create_sankey(
+    query_result: pd.DataFrame,
+    predefined_colors: Dict[str, str],
+    columns: Dict[str, str],
+    viz_config: Dict[str, Union[int, float]],
+    title: str = "unnamed",
+) -> Figure:
+    """
+    Creates a Sankey diagram based on input query_result
+    , which must contain source, target, value columns
```
Comment (bh2smith) [inline dune_client/viz/graphs.py:24]: Nit: Something about having a comma at the start of a line makes me feel uneasy
```suggestion
    Creates a Sankey diagram based on input query_result,
    which must contain source, target, value columns
```

**Review by bh2smith — APPROVED** [review]
I ran `make check` and black formatter just fixed your issues? Have you tried using the make file? 

I found some type issues as well and took care of them. These two changes are each in their own commit if you want to see what the relevant changes were for each. You should be good to go now.

Unfortunately its not going to be an easy task to add type annotations to these ignored projects so we will have to live with ignore statements for now.

Everything looks alright to me although I have not explicitly tried the functionality. You have my approval, but it might not hurt to get someone a Dune approval as well.

Thanks for your contribution!

File: tests/unit/test_viz_sankey.py (hunk 4 above)
Comment (agaperste) [reply tests/unit/test_viz_sankey.py]: got it makes sense, changed!

**Comment by agaperste** [comment]
@bh2smith Yes I ran `make check` in local but it was strangely not giving me errors 
And thank you so much for helping with this PR! 🙏

**Comment by agaperste** [comment]
@msf When you get a chance, could you take a look at this PR? First time merging into the repo so wanted to be sure :) Thanks in advance! 🙏

**Comment by bh2smith** [comment]
I'm excited to see this in action. I might even use it!

**Pull request merged**